x/evidence/types/msgs.go

```diff
@@ -11,38 +17,71 @@ const (
 )
 
 var (
-	_ sdk.Msg = MsgSubmitEvidenceBase{}
+	_ sdk.Msg                       = MsgSubmitEvidence{}
+	_ types.UnpackInterfacesMessage = MsgSubmitEvidence{}
+	_ exported.MsgSubmitEvidence    = MsgSubmitEvidence{}
 )
 
-// NewMsgSubmitEvidenceBase returns a new MsgSubmitEvidenceBase with a signer/submitter.
-// Note, the MsgSubmitEvidenceBase is not to be used as an actual message, but
-// rather to be extended with Evidence.
-func NewMsgSubmitEvidenceBase(s sdk.AccAddress) MsgSubmitEvidenceBase {
-	return MsgSubmitEvidenceBase{Submitter: s}
+// NewMsgSubmitEvidence returns a new MsgSubmitEvidence with a signer/submitter.
+func NewMsgSubmitEvidence(s sdk.AccAddress, evi exported.Evidence) (MsgSubmitEvidence, error) {
+	msg, ok := evi.(proto.Message)
+	if !ok {
+		return MsgSubmitEvidence{}, fmt.Errorf("cannot proto marshal %T", evi)
+	}
+	any, err := types.NewAnyWithValue(msg)
+	if err != nil {
+		return MsgSubmitEvidence{Submitter: s}, err
+	}
+	return MsgSubmitEvidence{Submitter: s, Evidence: any}, nil
 }
 
-// Route returns the MsgSubmitEvidenceBase's route.
-func (m MsgSubmitEvidenceBase) Route() string { return RouterKey }
+// Route returns the MsgSubmitEvidence's route.
+func (m MsgSubmitEvidence) Route() string { return RouterKey }
 
-// Type returns the MsgSubmitEvidenceBase's type.
-func (m MsgSubmitEvidenceBase) Type() string { return TypeMsgSubmitEvidence }
+// Type returns the MsgSubmitEvidence's type.
+func (m MsgSubmitEvidence) Type() string { return TypeMsgSubmitEvidence }
 
-// ValidateBasic performs basic (non-state-dependant) validation on a MsgSubmitEvidenceBase.
-func (m MsgSubmitEvidenceBase) ValidateBasic() error {
+// ValidateBasic performs basic (non-state-dependant) validation on a MsgSubmitEvidence.
+func (m MsgSubmitEvidence) ValidateBasic() error {
 	if m.Submitter.Empty() {
 		return sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, m.Submitter.String())
 	}
 
+	evi := m.GetEvidence()
+	if evi == nil {
+		return sdkerrors.Wrap(ErrInvalidEvidence, "missing evidence")
+	}
+	if err := evi.ValidateBasic(); err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // GetSignBytes returns the raw bytes a signer is expected to sign when submitting
-// a MsgSubmitEvidenceBase message.
-func (m MsgSubmitEvidenceBase) GetSignBytes() []byte {
+// a MsgSubmitEvidence message.
+func (m MsgSubmitEvidence) GetSignBytes() []byte {
 	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(m))
 }
 
-// GetSigners returns the single expected signer for a MsgSubmitEvidenceBase.
-func (m MsgSubmitEvidenceBase) GetSigners() []sdk.AccAddress {
+// GetSigners returns the single expected signer for a MsgSubmitEvidence.
+func (m MsgSubmitEvidence) GetSigners() []sdk.AccAddress {
 	return []sdk.AccAddress{m.Submitter}
 }
+
+func (m MsgSubmitEvidence) GetEvidence() exported.Evidence {
+	evi, ok := m.Evidence.GetCachedValue().(exported.Evidence)
```

Comment:
Is it guaranteed to be cached anytime this is called?

Reply:
No, it's only cached if:
* the client is calling this and the message was never sent over the wire
* `UnpackInterfaces` had been called after `Unmarshal` (which is the default behavior of `ProtoCodec.Unmarshal` but it would need to be implemented on `Tx` as well to unpack the `Msg`s)

So things should be fine if the default `ProtoCodec` and `Tx` are used. But if unpacking doesn't happen, it could be `nil`